### PR TITLE
feat(apigateway): rate limited API key

### LINKED ISSUE #6405: [Api-Gateway] Construct to allow quotas and throttling per api key
A construct which allows you to create a single api key with specified quotas and throttling limits. 

### Use Case
In cases when we need to configure throttling and quotas etc., at the api key level, we need to  
1. Create usage plan
2. Create an api key
3. Link the two (or specify the api key while creating the usage plan)

It would be nice to have a construct which abstracts the construction and linking of the api key and usage plan. 

This way, the user can just specify the quotas, throttling limits etc for the api key and not worry about usage plans and linking the two. 

### Proposed Solution

Adding a construct which allows you to create a single api key with specified quotas and throttling limits.

Internally, it would create a single usage plan with the specified configuration, along with a api key and link the two. 

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by vaibhav-walia**
## Commit Message
feat(apigateway): rate limited API key (#6509)

In cases when we need to configure throttling and quotas etc., at the api key level, we need to create a usage plan for the api key and link the two.
This commit introduces a construct which allows the user to create an api key and spicify rate limiting settings, while creating the api key and not worry about creating usage plan and linking the two.

fixes #6405
## End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * API keys are alphanumeric string values that you distribute to
+ * app developer customers to grant access to your API
+ */
+export interface IRateLimitedApiKey extends IResourceBase {
+  /**
+   * The API key ID.
+   * @attribute
+   */
+  readonly keyId: string;
+}
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: This interface is not needed. Use `IApiKey` in its place.

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+   * Overall throttle settings for the API.
+   * @default none
+   */
+  readonly throttle?: ThrottleSettings;
+}
+
+/**
+ * RateLimitedApiKey properties.
+ */
+export interface RateLimitedApiKeyProps extends ResourceOptions {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: If you extend `ApiKeyProps` instead, most of these properties don't need to be duplicated here.

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 3]
```diff
@@ -153,6 +153,34 @@ plan.addApiStage({
 });
 ```
 
+The following example shows how to use a rate limited api key :
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/README.md]: Can you add a paragraph here on what RateLimitedApiKey class is and when someone would use this?

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+   * The API key ID.
+   * @attribute
+   */
+  readonly keyId: string;
+}
+
+/**
+ * Rate limiting settings for the api key.
+ */
+export interface RateLimitingSettings {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Let's drop this interface and add all of these properties directly into the `RateLimitedApiKeyProps` interface.

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts (hunk 1 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Done!

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts (hunk 2 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Good Idea! Done!

File: packages/@aws-cdk/aws-apigateway/README.md (hunk 3 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/README.md]: Done!

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts (hunk 4 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Done!

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 5]
```diff
@@ -153,6 +153,36 @@ plan.addApiStage({
 });
 ```
 
+In scenarios where you need to create a single api key and configure rate limiting for it, you can use RateLimitedApiKey.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/README.md]:
```suggestion
In scenarios where you need to create a single api key and configure rate limiting for it, you can use `RateLimitedApiKey`.
```

File: packages/@aws-cdk/aws-apigateway/README.md [hunk 6]
```diff
@@ -153,6 +153,36 @@ plan.addApiStage({
...
+This construct lets you specify rate limiting properties which should be applied only to the api key being created.
+Internally, it automatically creates a usage plan with the specified rate limiting properties for this api key.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/README.md]:
```suggestion
The API key created has the specified rate limits, such as quota and throttles, applied.
```

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts [hunk 7]
```diff
@@ -20,6 +20,7 @@ export interface IApiKey extends IResourceBase {
  */
 export interface ApiKeyProps extends ResourceOptions {
   /**
+   * [disable-awslint:ref-via-interface]
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: Why is this needed? We generally don't encourage any new linter exclusions.

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      apiKeyName: props.apiKeyName,
+      customerId: props.customerId,
+      defaultCorsPreflightOptions: props.defaultCorsPreflightOptions,
+      defaultIntegration: props.defaultIntegration,
+      defaultMethodOptions: props.defaultMethodOptions,
+      description: props.description,
+      enabled: props.enabled,
+      generateDistinctId: props.generateDistinctId,
+      resources: props.resources
+    });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Can't we just do this instead?

```suggestion
const resource = new ApiKey(this, 'Resource', props);
```

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts (hunk 7 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: The build was failing after I made RateLimitedApiKeyProps extend ApiKeyProps :

```
error: [awslint:ref-via-interface:@aws-cdk/aws-apigateway.RateLimitedApiKeyProps.resources] API should use interface and not the concrete class (@aws-cdk/aws-apigateway.RestApi). If this is intentional, add "[disable-awslint:ref-via-interface]" to element's jsdoc
```
So I followed the suggestion and added ```[disable-awslint:ref-via-interface]``` to the jsdoc of ```resources``` property in ```ApiKeyProps```

Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: I see it now.

I believe this is occurring because this object is using `RestApi` instead of `IRestApi` as the type for `resources`. 
Can we try switching it to `IRestApi` and see if all tests still pass?

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import { Construct, Resource } from '@aws-cdk/core';
+import { ApiKey, ApiKeyProps, IApiKey } from './api-key';
+import { QuotaSettings, ThrottleSettings, UsagePlan, UsagePlanPerApiStage } from './usage-plan';
+
+/**
+ * RateLimitedApiKey properties.
+ */
+export interface RateLimitedApiKeyProps extends ApiKeyProps {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Missed something from last time. We usually don't extend one Props class from another Props class.

The following changes are needed here
- Rename `ApiKeyProps` to `ApiKeyOptions`
- Create a new empty `ApiKeyProps` class - 
  ```ts
  export interface ApiKeyProps extends ApiKeyOptions {
  }
  ```

Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: Not sure if I understand.

Is the suggestion to :
1. Rename ```ApiKeyProps``` defined in ```ApiKey```, to ```ApiKeyOptions```
2. Create an empty interface (```ApiKeyProps```) in ```RateLimitedApiKey``` extending ```ApiKeyOptions```
3. Use it like 
```typescript
export interface RateLimitedApiKeyProps extends ApiKeyProps {
```

If that's the case, I'm confused how it will help/is better

File: packages/@aws-cdk/aws-apigateway/lib/api-key.ts (hunk 7 above)
Comment (vaibhav-walia) [reply packages/@aws-cdk/aws-apigateway/lib/api-key.ts]: I tried doing that, but it resulted in compilation errors as ```renderStageKeys``` method expects ```resources``` to be of type ```RestApi``` and changing it to ```IRestApi``` is not possible as it doesn't have ```deploymentStage``` property needed by this method 
https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-apigateway/lib/api-key.ts#L90-L101

File: packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts (hunk 9 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/rate-limited-api-key.ts]: You're correct. I think I was looking for an optimization/future-proofing whose value is unclear yet. Ignore this.
